internal/auth0/role/resource_permissions.go

```diff
@@ -0,0 +1,174 @@
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Name of resource server that the permission is associated with.",
+						},
+					},
+				},
+			},
+		},
+		CreateContext: upsertRolePermissions,
+		UpdateContext: upsertRolePermissions,
+		ReadContext:   readRolePermissions,
+		DeleteContext: deleteRolePermissions,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Description: "With this resource, you can manage role permissions (1-many).",
+	}
+}
+
+func upsertRolePermissions(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	api := meta.(*config.Config).GetAPI()
+	mutex := meta.(*config.Config).GetMutex()
+
+	roleID := data.Get("role_id").(string)
+
+	if !data.HasChange("permissions") {
+		return nil
+	}
+
+	mutex.Lock(roleID)
+	defer mutex.Unlock(roleID)
+
+	toAdd, toRemove := value.Difference(data, "permissions")
+
+	var addPermissions []*management.Permission
+	for _, addPermission := range toAdd {
+		permission := addPermission.(map[string]interface{})
+		addPermissions = append(addPermissions, &management.Permission{
+			Name:                     auth0.String(permission["name"].(string)),
+			ResourceServerIdentifier: auth0.String(permission["resource_server_identifier"].(string)),
+		})
+	}
+
+	if len(addPermissions) > 0 {
+		if err := api.Role.AssociatePermissions(roleID, addPermissions); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	var rmPermissions []*management.Permission
+	for _, rmPermission := range toRemove {
+		permission := rmPermission.(map[string]interface{})
+		rmPermissions = append(rmPermissions, &management.Permission{
+			Name:                     auth0.String(permission["name"].(string)),
+			ResourceServerIdentifier: auth0.String(permission["resource_server_identifier"].(string)),
+		})
+	}
+
+	if len(rmPermissions) > 0 {
+		if err := api.Role.RemovePermissions(roleID, rmPermissions); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	data.SetId(roleID)
+
+	return readRolePermissions(ctx, data, meta)
+}
+
+func readRolePermissions(_ context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	api := meta.(*config.Config).GetAPI()
+
+	roleID := data.Get("role_id").(string)
+
+	permissions, err := api.Role.Permissions(roleID)
+	if err != nil {
+		if mErr, ok := err.(management.Error); ok && mErr.Status() == http.StatusNotFound {
+			data.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	err = data.Set("permissions", flattenRolePermissions(permissions.Permissions))
+
+	return diag.FromErr(err)
+}
+
+func deleteRolePermissions(_ context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	api := meta.(*config.Config).GetAPI()
+	mutex := meta.(*config.Config).GetMutex()
+
+	roleID := data.Get("role_id").(string)
+
+	mutex.Lock(roleID)
+	defer mutex.Unlock(roleID)
+
+	permissions, err := api.Role.Permissions(roleID)
```

Comment:
Hm... 🤔 This will remove all the permissions, not just the ones assigned through this resource. *I think* it should be okay, however in the auth0_user_roles we do 

```
userRolesToRemove := data.Get("roles").(*schema.Set).List()
	var rmRoles []*management.Role
	for _, rmRole := range userRolesToRemove {
		role := &management.Role{ID: auth0.String(rmRole.(string))}
		rmRoles = append(rmRoles, role)
	}
```

The two approaches technically should be exactly the same as we should manage all the permissions through this and have nothing left on the API. 

However, there is one edge case where:

1. You attach permissions to a role. (but the role has actually more permissions set on the API)
2. You do a read in another terraform plan and you're gonna see the diff between what you have in the config and what's on remote.
3. If you don't add those extra permissions to the config and immediately do a terraform destroy, with the logic we have now it will remove even the others that are untracked, whereas the solution above to fetch them again from the config will only remove the ones that we set in the config and leave the others that were untracked and shown in the diff.

What do you think we should do?

Reply:
While I think it would be an anti-pattern for folks to adopt that workflow, it's a safer and low-cost to only delete the ones within management purview.